### PR TITLE
Implement Manifest Manager

### PR DESCRIPTION
**Pull request by fraxken**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -132,6 +132,7 @@ Click on one of the links to access the documentation of the workspace:
 | --- | --- |
 | tarball | [@nodesecure/sec-literal](./workspaces/tarball) |
 | tree-walker | [@nodesecure/tree-walker](./workspaces/tree-walker) |
+| mama | [@nodesecure/tree-walker](./workspaces/mama) |
```
Comment (PierreDemailly) [inline README.md:135]:
```suggestion
| mama | [@nodesecure/mama](./workspaces/mama) |
```

File: workspaces/mama/package.json [hunk 2]
```diff
@@ -0,0 +1,38 @@
+{
+  "name": "@nodesecure/mama",
+  "version": "1.0.0",
+  "description": "Manifest Manager",
+  "type": "module",
+  "exports": "./dist/index.js",
+  "types": "./dist/index.d.ts",
+  "scripts": {
+    "build": "tsc -b",
```
Comment (PierreDemailly) [inline workspaces/mama/package.json:9]:
```suggestion
    "build": "tsc -b",
    "prepublishOnly": "npm run build",
```

File: workspaces/mama/src/index.ts [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+        packageStr
+      ) as PackageJSON | WorkspacesPackageJSON;
+
+      return new ManifestManager(
+        packageJSON
+      );
+    }
+    catch (cause) {
+      throw new Error(
+        `Failed to parse package.json located at: ${packageLocation} (see the cause for additional information).`,
```
Comment (PierreDemailly) [inline workspaces/mama/src/index.ts:153]:
```suggestion
        `Failed to parse package.json located at: ${packageLocation}`,
```
I'm ok to keep the `(see the cause for additional information)` but it feel unecessary to me. Anyway the cause will be shown and `cause` should be obvious for our library users target, wdyt ?

File: workspaces/mama/test/ManifestManager.spec.ts [hunk 4]
```diff
@@ -0,0 +1,573 @@
...
+      catch (error) {
+        assert.strictEqual(error.name, "Error");
+        assert.strictEqual(
+          error.message,
+          `Failed to parse package.json located at: ${expectedLocation} (see the cause for additional information).`
+        );
+
+        assert.ok("cause" in error);
+        assert.strictEqual(error.cause.name, "SyntaxError");
+      }
```
Comment (PierreDemailly) [inline workspaces/mama/test/ManifestManager.spec.ts:123]: I guess its intended to not use `assert.rejects` here ?

File: workspaces/tarball/src/utils/analyzeDependencies.ts [hunk 5]
```diff
@@ -1,26 +1,74 @@
...
+
 // Import Third-party Dependencies
-import difference from "lodash.difference";
-// @ts-ignore
-import builtins from "builtins";
+import { ManifestManager } from "@nodesecure/mama";
 
 // Import Internal Dependencies
 import { getPackageName } from "./getPackageName.js";
+import type { NodeImport } from "@nodesecure/npm-types";
```
Comment (PierreDemailly) [inline workspaces/tarball/src/utils/analyzeDependencies.ts:9]: move to 3rd party ?

File: workspaces/tarball/src/utils/analyzeDependencies.ts [hunk 6]
```diff
@@ -30,33 +78,38 @@ export interface analyzeDependenciesResult {
...
-    .map((name) => (packageDeps.includes(name) ? name : getPackageName(name)))
-    .filter((name) => !name.startsWith("."))
-    .filter((name) => !isNodeCoreModule(name))
-    .filter((name) => !packageDevDeps.includes(name))
+  const thirdPartyDependencies = sourceDependencies
+    .map((name) => (dependencies.includes(name) ? name : getPackageName(name)))
+    .filter((name) => !isFile(name))
+    .filter((name) => !isCoreModule(name))
+    .filter((name) => !devDependencies.includes(name))
     .filter((name) => !tryDependencies.has(name));
```
Comment (PierreDemailly) [inline workspaces/tarball/src/utils/analyzeDependencies.ts:103]: I think we can use a flatMap here and remove each filter so we loop only once, wdyt ?

Something like this:
```ts
const thirdPartyDependencies = sourceDependencies
  .flatMap((sourceName) => {
    const name = dependencies.includes(sourceName) ? sourceName : getPackageName(sourceName);

    if (isFile(name)) {
      return [];
    }

    if (isCoreModule(name)) {
      return [];
    }

    if (devDependencies.includes(name)) {
      return [];
    }

    if (tryDependencies.has(name)) {
      return [];
    }

    return name;
  });
```

File: workspaces/tarball/src/utils/analyzeDependencies.ts [hunk 7]
```diff
@@ -72,27 +125,38 @@ export function analyzeDependencies(
...
+) {
+  return name.startsWith(".") || path.extname(name) !== ""
+}
+
+function isCoreModule(
+  moduleName: string
+): boolean {
   const cleanModuleName = moduleName.startsWith("node:") ? moduleName.slice(5) : moduleName;
 
   // Note: We need to also check moduleName because builtins package only return true for 'node:test'.
```
Comment (PierreDemailly) [inline workspaces/tarball/src/utils/analyzeDependencies.ts:143]: We can remove/rewrite this comment as we don't use builtins package anymore

File: workspaces/mama/test/ManifestManager.spec.ts (hunk 4 above)
Comment (fraxken) [reply workspaces/mama/test/ManifestManager.spec.ts:123]: Yes, doesn't work well for cause

File: workspaces/tarball/src/utils/analyzeDependencies.ts (hunk 7 above)
Comment (fraxken) [reply workspaces/tarball/src/utils/analyzeDependencies.ts:143]: No that's still valid, builtins here is not about the package name 😉

Comment (PierreDemailly) [reply workspaces/tarball/src/utils/analyzeDependencies.ts:143]: oh, ok ^^

**Review by PierreDemailly — APPROVED** [review]

**Pull request merged**